import libraries


In [1]:
import numpy as np
import math
import pandas as pd
import seaborn as sns
import matplotlib as plt

import file

In [2]:

xlsx_file_path = 'C:/Users/Lenovo/Desktop/ironhack/Week_5/Mid_bootcamp_project/data/raw/project_file.xlsx'

sheet_name = 'Data' 

data = pd.read_excel(xlsx_file_path, sheet_name=sheet_name)


visualize data

In [3]:
data.shape

(28259, 122)

In [4]:
data.head()

Country Code Country_ID                     Product_ID Product Code  \
0             8    Albania  Roundwood (wood in the rough)            1   
1             8    Albania  Roundwood (wood in the rough)            1   
2             8    Albania  Roundwood (wood in the rough)            1   
3             8    Albania  Roundwood (wood in the rough)            1   
4             8    Albania  Roundwood (wood in the rough)            1   

       Flow      Unit   1964D 1964S   1965D 1965S  ...    2017D 2017S  \
0   EXPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    26.67     C   
1   EXPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1676.93     C   
2   IMPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    30.47     C   
3   IMPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1648.96     C   
4  REMOVALS   1000 m3  2025.0     C  2060.0     C  ...  1230.01     C   

     2018D 2018S    2019D 2019S    2020D 2020S    2021D 2021S  
0    15.35     C    14.02     C    11.80     C    14.26     C  
1  1907.49     C  1804.94     C  1692.82     C  2214.79     C  
2    23.80     C    18.46     C    39.04     C    23.37     C  
3  2325.06     C  1548.85     C  3807.80     C  2199.58     C  
4   905.34     C   917.84     C   874.45     C  1021.91     C  

[5 rows x 122 columns]

In [5]:
data.tail()

Country Code             Country_ID  \
28254           891  Serbia and Montenegro   
28255           891  Serbia and Montenegro   
28256           891  Serbia and Montenegro   
28257           891  Serbia and Montenegro   
28258           891  Serbia and Montenegro   

                                              Product_ID Product Code  \
28254                   Non-coniferous sawnwood of birch   6.NC.Birch   
28255  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28256  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28257  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28258  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   

          Flow      Unit  1964D 1964S  1965D 1965S  ...  2017D 2017S  2018D  \
28254  IMPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28255  EXPORTS   1000 m3    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28256  EXPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28257  IMPORTS   1000 m3    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28258  IMPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   

      2018S  2019D 2019S  2020D 2020S  2021D 2021S  
28254   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28255   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28256   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28257   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28258   NaN    NaN   NaN    NaN   NaN    NaN   NaN  

[5 rows x 122 columns]

DATA CLEANING AND WRANGLING


In [6]:
#get columns names
data.columns

Index(['Country Code', 'Country_ID', 'Product_ID', 'Product Code', 'Flow',
       'Unit', '1964D', '1964S', '1965D', '1965S',
       ...
       '2017D', '2017S', '2018D', '2018S', '2019D', '2019S', '2020D', '2020S',
       '2021D', '2021S'],
      dtype='object', length=122)

some definitions are given as

UNECE/FAO TIMBER database, 1964-2021, as of March 2022. 

Data point (yearD), status (yearS). 

Status
C=calculated (sum of subitems) -
E1=technical estimate (based on similar data) 
E2=analyst estimate (based on outside source) 
E3=estimate based on official aggregate - R=repeated - N=national estimate - blank=official

In [7]:
#define function  to get unique values in a column

def unique_values(column):
    return data[column].unique()


In [8]:
unique_values('Country_ID')

array(['Albania', 'Andorra', 'Azerbaijan', 'Austria', 'Armenia',
       'Belgium', 'Belgium & Luxembourg', 'Bosnia and Herzegovina',
       'Bulgaria', 'Belarus', 'Canada', 'Croatia', 'Cyprus',
       'Former CSFR', 'Czechia', 'Denmark', 'Estonia', 'Finland',
       'France', 'Georgia', 'Germany',
       'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Kazakhstan',
       'Kyrgyzstan', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg',
       'Malta', 'Republic of Moldova', 'Montenegro', 'Netherlands',
       'Norway', 'Poland', 'Portugal', 'Romania', 'Russian Federation',
       'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',
       'Tajikistan', 'Turkey', 'Turkmenistan', 'Ukraine',
       'North Macedonia', 'Former Soviet Union', 'United Kingdom',
       'United States', 'Uzbekistan', 'Former SFRY',
       'Serbia and Montenegro'], dtype=object)

DEALING WITH COUNTRIES


for convineience of analysis, our focus will be on classsic european nation 
Russia is problematic for our analysis as most of her older data are contained in the soviet union, and unless we aggregate the data of other constuitent states that composes the defunct USSR (see link, https://en.wikipedia.org/wiki/Post-Soviet_states), we would introduce errors if we are to include the data from russia .
so for convinient sake we will exclude Russia , future project might require using future estimates from russia and other post-soviet states to determine 
The post soviet states include
 1.  Armenia •
 2.  Azerbaijan • 
 3.  Belarus •
 4.  Estonia •
 5.  Georgia • 
 6.  Kazakhstan • 
 7.  Kyrgyzstan • 
 8.  Latvia • 
 9.  Lithuania • 
 10. Moldova • 
 11. Russia • 
 12. Tajikistan •
 13. Turkmenistan • 
 14. Ukraine 
 15. Uzbekistan

for other defunct states like serbia and montenegro, Czech and Slovak Federative Republic (CSFR),The Socialist Federal Republic of Yugoslavia (SFRY) we will also drop the countries for this analysis, a future possibilty  might require using them
so by that , we will drop mordern states like 


and also we will create a seperate data set that contain mordern data where only mordern states are included and use it answer question 2 . 

germany was split before 1993 to west(Germany, Former Federal Republic of) and east (Former German Democratic Republic')

we will concatenate both rows in a new data set and include as simply germany. Historically , the two partitioned german state has not lost any territory before and after reunification. So we won´t introduce any error to our data

In [9]:
unique_values('Product_ID')

array(['Roundwood (wood in the rough)', 'Coniferous roundwood',
       'Non-Coniferous roundwood',
       'Wood fuel, including wood for charcoal', 'Coniferous wood fuel',
       'Non-Coniferous wood fuel', 'Industrial roundwood',
       'Coniferous industrial roundwood',
       'Non-Coniferous industrial roundwood',
       'Tropical industrial roundwood',
       'Non-Coniferous non-tropical industrial roundwood',
       'Sawlogs and veneer logs', 'Coniferous sawlogs and veneer logs',
       'Non-Coniferous sawlogs and veneer logs',
       'Pulpwood (round & split)', 'Coniferous pulpwood',
       'Non-Coniferous Pulpwood', 'Other industrial roundwood',
       'Coniferous other industrial roundwood',
       'Non-Coniferous other industrial roundwood', 'Wood charcoal',
       'Wood residues, chips and particles', 'Chips and particles',
       'Wood residues', 'Recovered post-consumer wood',
       'Wood pellets and other agglomerates of wood', 'Wood pellets',
       'Other agglomerates o

In [10]:
#number of unique values in a column
def unique_counts(column):
    return data[column].nunique()

unique_counts('Product_ID')


130

Most efforts wil be centered around product_id column, we will reduce the variability by removing the sub-categories that tries to distinguish between coniferous, non-coniferous and others. For a study on forest ecosystem dynamism, that might be a useful categorization, but not for this study. 

Also all categories of paper productrs and packaging product will be categorized as paper and packaging
At the end we will have only paper and packaging.
All kinds of pulp will be joined, so it won´t matter if it´s kraft, sulfite , bleached or unbleached, virgin or recovered fibre 

All engineering boards (wood based panels) like veneer, plywood, osb, hardboard, MDF/HDF, particleboard,otherboards etc will be grouped under engineering board.

wood pellets and other agglomerate will be grouped together
wood residues, chips and pellets will be gropued together too
all pulpwood will be grouped 
.
At the end we will significantly reduce product category from 130 to less than 20.

In [11]:
#lets see what a row that has Roundwood (wood in the rough) looks like
data[data['Product_ID'] == 'Roundwood (wood in the rough)'].head()

Country Code Country_ID                     Product_ID Product Code  \
0             8    Albania  Roundwood (wood in the rough)            1   
1             8    Albania  Roundwood (wood in the rough)            1   
2             8    Albania  Roundwood (wood in the rough)            1   
3             8    Albania  Roundwood (wood in the rough)            1   
4             8    Albania  Roundwood (wood in the rough)            1   

       Flow      Unit   1964D 1964S   1965D 1965S  ...    2017D 2017S  \
0   EXPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    26.67     C   
1   EXPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1676.93     C   
2   IMPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    30.47     C   
3   IMPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1648.96     C   
4  REMOVALS   1000 m3  2025.0     C  2060.0     C  ...  1230.01     C   

     2018D 2018S    2019D 2019S    2020D 2020S    2021D 2021S  
0    15.35     C    14.02     C    11.80     C    14.26     C  
1  1907.49     C  1804.94     C  1692.82     C  2214.79     C  
2    23.80     C    18.46     C    39.04     C    23.37     C  
3  2325.06     C  1548.85     C  3807.80     C  2199.58     C  
4   905.34     C   917.84     C   874.45     C  1021.91     C  

[5 rows x 122 columns]

CLEANING FLOW

In [12]:
unique_values('Flow')

array(['EXPORTS', 'IMPORTS', 'REMOVALS', 'PRODUCTION'], dtype=object)

To answer our first question; REMOVALS will be regarded as harvest,
trade deficit will be looked into comparing import and exports 
#Production cannot be a good determinant and that might also include locally and imported raw materials #

In [13]:
unique_values('2015S')

array(['C', 'R', nan, 'E1', 'E2', 'N', 'E3'], dtype=object)


Due to the nature of our analysis, the yearStatus column will not be put into consideration. 
Status
C=calculated (sum of subitems) -
E1=technical estimate (based on similar data) 
E2=analyst estimate (based on outside source) 
E3=estimate based on official aggregate - R=repeated - N=national estimate - blank=official

so we are more concerned with the value than the way it was calculated ,as we assume that the calculation method used is reliable or at least approximate to actual value

CLEANING THE DATA

OBJECTIVE 1
-clean up product 
how- define functions to  identify keywords in product_id and group them on a new column named named product
at the end we will have products reduced to 
-roundwood
-veneer logs/sawlogs
-pulpwood
-wood charcoal
wood residue /chips&particle
wood pellets &other agglomerates
-sawnwood
-wood based panels
-wood pulp
-paper and packaging 
-processed sawn wood
-other wood product
A lot of products will be classified as others , in a future analysis, these products will be analyzed on individual merit


In [ ]:
#,define function to create a new column which will be made by checking the character in a column and assigninga string value to that character which will form the new column
def new_column(char):
    '''
    This function creates a new column based on the character in a column
    param char: character in a column
    output: new column with string values
    '''
    new_column = []
    for i in data[char]:
        if i == 'roundwood':
            new_column.append('roundwood')
        elif i == 'veneer':
            new_column.append('Export')
        else:
            new_column.append('Re-Export')
